File: examples/python/tenant_rate_limit_deadline.py
"""Tenant-aware OpenAI-compatible client with a global deadline.

This example keeps retry and fallback budgets explicit. In production, replace
TenantLimiter with Redis, a queue, or gateway-side rate limiting so limits are
shared across service instances.
"""
from __future__ import annotations

import json
import os
import time
import uuid
from dataclasses import dataclass
from typing import Iterable

from openai import APIConnectionError, APIStatusError, APITimeoutError, OpenAI, RateLimitError


@dataclass(frozen=True)
class Route:
    model: str
    cost_group: str
    timeout: float


ROUTES = {
    "support": [
        Route("claude-sonnet-4", "stable_official", 10),
        Route("gpt-4o-mini", "official_transfer", 8),
    ],
    "batch": [
        Route("gemini-2.5-flash", "welfare", 20),
        Route("gpt-4o-mini", "official_transfer", 12),
    ],
}

RETRYABLE = (APIConnectionError, APITimeoutError, RateLimitError)


class TenantLimiter:
    """Small-process demo; use a shared limiter for multi-instance services."""

    def __init__(self, per_tenant: int = 2) -> None:
        self.per_tenant = per_tenant
        self.active: dict[str, int] = {}

    def acquire(self, tenant_id: str) -> bool:
        current = self.active.get(tenant_id, 0)
        if current >= self.per_tenant:
            return False
        self.active[tenant_id] = current + 1
        return True

    def release(self, tenant_id: str) -> None:
        self.active[tenant_id] = max(self.active.get(tenant_id, 1) - 1, 0)


def log_event(**fields: object) -> None:
    print(json.dumps(fields, ensure_ascii=False, sort_keys=True))


def complete(*, client: OpenAI, limiter: TenantLimiter, tenant_id: str,
             scenario: str, messages: Iterable[dict[str, str]]) -> str:
    request_id = str(uuid.uuid4())
    if not limiter.acquire(tenant_id):
        log_event(event="tenant_rate_limited", request_id=request_id, tenant_id=tenant_id)
        raise RuntimeError("tenant concurrency limit reached")

    deadline = time.monotonic() + (25 if scenario == "support" else 90)
    try:
        for fallback_index, route in enumerate(ROUTES[scenario]):
            for attempt in range(1, 3):
                if time.monotonic() >= deadline:
                    break
                started = time.monotonic()
                try:
                    response = client.chat.completions.create(
                        model=route.model,
                        messages=list(messages),
                        timeout=min(route.timeout, max(deadline - time.monotonic(), 1)),
                    )
                    log_event(event="llm_call_ok", request_id=request_id,
                              tenant_id=tenant_id, scenario=scenario,
                              model=route.model, cost_group=route.cost_group,
                              fallback_index=fallback_index, attempt=attempt,
                              degraded=fallback_index > 0,
                              latency_ms=round((time.monotonic() - started) * 1000))
                    return response.choices[0].message.content or ""
                except RETRYABLE as exc:
                    log_event(event="llm_call_retryable", request_id=request_id,
                              tenant_id=tenant_id, model=route.model,
                              attempt=attempt, error_type=type(exc).__name__)
                    if attempt < 2:
                        time.sleep(min(2**attempt, 4))
                except APIStatusError as exc:
                    log_event(event="llm_call_status_error", request_id=request_id,
                              tenant_id=tenant_id, model=route.model,
                              status_code=exc.status_code)
                    if exc.status_code < 500:
                        raise
        raise RuntimeError(f"all routes failed: request_id={request_id}")
    finally:
        limiter.release(tenant_id)


if __name__ == "__main__":
    sdk = OpenAI(api_key=os.environ["VIRALAPI_API_KEY"],
                 base_url=os.getenv("VIRALAPI_BASE_URL", "https://viralapi.ai/v1"),
                 timeout=15, max_retries=0)
    print(complete(client=sdk, limiter=TenantLimiter(), tenant_id="demo",
                   scenario="support",
                   messages=[{"role": "user", "content": "Summarize this ticket."}]))

File: examples/node/tenant-rate-limit-deadline.mjs
import crypto from "node:crypto";
import OpenAI from "openai";

const client = new OpenAI({
  apiKey: process.env.VIRALAPI_API_KEY,
  baseURL: process.env.VIRALAPI_BASE_URL || "https://viralapi.ai/v1",
  timeout: Number(process.env.VIRALAPI_TIMEOUT_MS || 12000),
  maxRetries: 0,
});

const routes = {
  support: [
    { model: "claude-sonnet-4", group: "stable_official", timeoutMs: 10000 },
    { model: "gpt-4o-mini", group: "official_transfer", timeoutMs: 8000 },
  ],
  batch: [
    { model: "gemini-2.5-flash", group: "welfare", timeoutMs: 20000 },
    { model: "gpt-4o-mini", group: "official_transfer", timeoutMs: 12000 },
  ],
};

const sleep = (ms) => new Promise((resolve) => setTimeout(resolve, ms));
const retryable = (status) => [408, 425, 429, 500, 502, 503, 504, "ETIMEDOUT"].includes(status);

export async function complete({ tenantId, scenario, messages }) {
  const requestId = crypto.randomUUID();
  const deadline = Date.now() + (scenario === "support" ? 25000 : 90000);

  for (const [fallbackIndex, route] of routes[scenario].entries()) {
    for (let attempt = 1; attempt <= 2 && Date.now() < deadline; attempt += 1) {
      try {
        const response = await client.chat.completions.create({
          model: route.model,
          messages,
          timeout: Math.min(route.timeoutMs, Math.max(deadline - Date.now(), 1000)),
        });
        console.log(JSON.stringify({ event: "llm_call_ok", request_id: requestId,
          tenant_id: tenantId, scenario, model: route.model, cost_group: route.group,
          fallback_index: fallbackIndex, attempt, degraded: fallbackIndex > 0 }));
        return response.choices[0].message.content;
      } catch (error) {
        const status = error.status || error.code;
        console.log(JSON.stringify({ event: "llm_call_error", request_id: requestId,
          tenant_id: tenantId, scenario, model: route.model, cost_group: route.group,
          fallback_index: fallbackIndex, attempt, status }));
        if (!retryable(status)) throw error;
        if (attempt < 2) await sleep(Math.min(1000 * 2 ** attempt, 4000));
      }
    }
  }
  throw new Error(`llm_deadline_exceeded request_id=${requestId}`);
}
